grid-client/deployer/network_deployer.go

```diff
@@ -52,36 +56,139 @@ func (d *NetworkDeployer) Validate(ctx context.Context, znet *workloads.ZNet) er
 }
 
 // GenerateVersionlessDeployments generates deployments for network deployer without versions.
-// usedPorts can be used to exclude some ports from being assigned to networks
-func (d *NetworkDeployer) GenerateVersionlessDeployments(ctx context.Context, znet *workloads.ZNet, usedPorts map[uint32][]uint16) (map[uint32]gridtypes.Deployment, error) {
+func (d *NetworkDeployer) GenerateVersionlessDeployments(ctx context.Context, znets []*workloads.ZNet) (map[uint32][]gridtypes.Deployment, error) {
+	deployments := make(map[uint32][]gridtypes.Deployment)
+	endpoints := make(map[uint32]net.IP)
+	nodeUsedPorts := make(map[uint32][]uint16)
+	allNodes := make(map[uint32]struct{})
+	var multiErr error
+
+	for _, znet := range znets {
+		for _, node := range znet.Nodes {
+			allNodes[node] = struct{}{}
+		}
+		if znet.PublicNodeID != 0 {
+			allNodes[znet.PublicNodeID] = struct{}{}
+		}
+	}
+
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+
+	for nodeID := range allNodes {
+		wg.Add(1)
+		go func(nodeID uint32) {
+			defer wg.Done()
+			endpoint, usedPorts, err := d.getNodeEndpointAndPorts(ctx, nodeID)
+			mu.Lock()
+			defer mu.Unlock()
+			if err != nil {
+				multiErr = multierror.Append(multiErr, err)
+				return
+			}
+			endpoints[nodeID] = endpoint
+			nodeUsedPorts[nodeID] = usedPorts
+		}(nodeID)
+
+	}
+	wg.Wait()
+
+	var publicNode uint32
+	var err error
+	for nodeID := range allNodes {
+		if endpoints[nodeID] != nil && endpoints[nodeID].To4() != nil {
+			publicNode = nodeID
+			break
+		}
+	}
+
+	if publicNode == 0 && needPublicNode(znets) {
+		publicNode, err = GetPublicNode(ctx, *d.tfPluginClient, nil)
+		if err != nil {
+			multiErr = multierror.Append(
+				multiErr,
+				fmt.Errorf("failed to get public node: %w", err),
+			)
+		}
+	}
+	if _, ok := endpoints[publicNode]; !ok && publicNode != 0 {
+		endpoint, usedPorts, err := d.getNodeEndpointAndPorts(ctx, publicNode)
+		if err != nil {
+			multiErr = multierror.Append(multiErr, err)
+		} else {
+			endpoints[publicNode] = endpoint
+			nodeUsedPorts[publicNode] = usedPorts
+		}
+	}
+
+	for _, znet := range znets {
```

Comment:
should we use concurrency here? no?

Reply:
these are all local processing, no requests are done over network, so they will end up sequential anyway since they are not blocking